### PR TITLE
Add polars plugin

### PR DESCRIPTION
**Pull request by RobinKa**
# TL;DR
Polars is a fast dataframe library.

~~This PR adds a type transformer for the `polars.DataFrame` type.~~
~~The plugin is mostly a copy of the Modin plugin with `polars.DataFrame` instead.~~

Adds support for using `polars.DataFrame` with structured datasets. Following https://docs.flyte.org/projects/cookbook/en/latest/auto/core/type_system/structured_dataset.html and the Spark plugin.

Tested with polars 0.8.27 (when `.transpose()` was added, which is used in the deck code) and 0.13.44/45. Versions before 0.13.12 require `to_parquet` instead of `write_parquet`, which is handled in the plugin code. 

## Type
 - [ ] Bug Fix
 - [ ] Feature
 - [x] Plugin

## Are all requirements met?

 - [x] Code completed
 - [ ] Smoke tested
 - [x] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

### CODE REVIEW

File: plugins/flytekit-polars/flytekitplugins/polars/schema.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    ) -> T:
+        if not (lv and lv.scalar and lv.scalar.schema):
+            return pl.DataFrame()
+        local_dir = ctx.file_access.get_random_local_directory()
+        ctx.file_access.download_directory(lv.scalar.schema.uri, local_dir)
+        r = PolarsSchemaReader(from_path=local_dir, cols=None, fmt=SchemaFormat.PARQUET)
+        return r.all()
+
+
+SchemaEngine.register_handler(
```
Comment (pingsutw) [inline plugins/flytekit-polars/flytekitplugins/polars/schema.py:112]: @RobinKa Thank you for your work. We may deprecate the scheme engine in the future. could you register this handler to `StructuredDatasetTransformerEngine` instead? Check this example https://github.com/flyteorg/flytekit/blob/master/plugins/flytekit-spark/flytekitplugins/spark/sd_transformers.py

Comment (RobinKa) [reply plugins/flytekit-polars/flytekitplugins/polars/schema.py:112]: Do we still want the schema stuff at all or do we only want the StructuredDataset ones?

File: flytekit/types/structured/structured_dataset.py [hunk 2]
```diff
@@ -647,6 +649,8 @@ def to_html(self, ctx: FlyteContext, python_val: typing.Any, expected_python_typ
             return pd.DataFrame(df).describe().to_html()
         elif importlib.util.find_spec("pyspark") is not None and isinstance(df, pyspark.sql.DataFrame):
             return pd.DataFrame(df.schema, columns=["StructField"]).to_html()
+        elif importlib.util.find_spec("polars") is not None and isinstance(df, pl.DataFrame):
+            return df.to_pandas().describe().to_html()
```
Comment (RobinKa) [inline flytekit/types/structured/structured_dataset.py:653]: This is pretty inefficient as it doubles the memory usage (making a copy of the dataframe). Polars itself doesn't have html rendering so I guess we need to write an html renderer for it?

File: plugins/flytekit-polars/flytekitplugins/polars/sd_transformers.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+    def encode(
+        self,
+        ctx: FlyteContext,
+        structured_dataset: StructuredDataset,
+        structured_dataset_type: StructuredDatasetType,
+    ) -> literals.StructuredDataset:
+        df = typing.cast(pl.DataFrame, structured_dataset.dataframe)
+
+        local_dir = typing.cast(str, structured_dataset.uri) or ctx.file_access.get_random_local_directory()
```
Comment (RobinKa) [inline plugins/flytekit-polars/flytekitplugins/polars/sd_transformers.py:32]: I'm not sure about this part. In the numpy example [here](https://docs.flyte.org/projects/cookbook/en/latest/auto/core/type_system/structured_dataset.html) it only has `ctx.file_access.get_random_local_directory()` but the Spark plugin had this extra `typing.cast` part.

File: flytekit/types/structured/structured_dataset.py (hunk 2 above)
Comment (RobinKa) [reply flytekit/types/structured/structured_dataset.py:653]: Ah I see for Spark we only render the schema, should do the same.

File: plugins/flytekit-polars/flytekitplugins/polars/sd_transformers.py (hunk 3 above)
Comment (pingsutw) [inline plugins/flytekit-polars/flytekitplugins/polars/sd_transformers.py:32]:
```suggestion
        local_dir = ctx.file_access.get_random_local_directory()
```

File: plugins/flytekit-polars/flytekitplugins/polars/sd_transformers.py [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+        local_path = f"{local_dir}/00000"
+
+        # Polars 0.13.12 deprecated to_parquet in favor of write_parquet
+        if hasattr(df, "write_parquet"):
+            df.write_parquet(local_path)
+        else:
+            df.to_parquet(local_path)
+        remote_dir = ctx.file_access.get_random_remote_directory()
```
Comment (pingsutw) [inline plugins/flytekit-polars/flytekitplugins/polars/sd_transformers.py:40]:
```suggestion
        remote_dir = typing.cast(str, structured_dataset.uri) or ctx.file_access.get_random_remote_directory()
```

File: plugins/flytekit-polars/flytekitplugins/polars/sd_transformers.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+        return pl.read_parquet(path)
+
+
+for protocol in [LOCAL, S3]:
+    StructuredDatasetTransformerEngine.register(
+        PolarsDataFrameToParquetEncodingHandler(protocol), default_for_type=True
+    )
+    StructuredDatasetTransformerEngine.register(
+        ParquetToPolarsDataFrameDecodingHandler(protocol), default_for_type=True
+    )
```
Comment (pingsutw) [inline plugins/flytekit-polars/flytekitplugins/polars/sd_transformers.py:70]: Need to register the handler for GCS as well, but `default_for_type` is false
```python
StructuredDatasetTransformerEngine.register(PolarsDataFrameToParquetEncodingHandler(GCS), default_for_type=False)
StructuredDatasetTransformerEngine.register(ParquetToPolarsDataFrameDecodingHandler(GCS), default_for_type=False)
```

File: plugins/flytekit-polars/flytekitplugins/polars/schema.py (hunk 1 above)
Comment (pingsutw) [reply plugins/flytekit-polars/flytekitplugins/polars/schema.py]: we only want the StructuredDataset ones

**Review by pingsutw — COMMENTED** [review]
Others LGTM, thank you

**Review by pingsutw — APPROVED** [review]

**Review by eapolinario — APPROVED** [review]
This is amazing.
